### PR TITLE
Add auto-healing tool for gateways in k8s cluster

### PR DESCRIPTION
**Pull request by vladiskuz**
This is standalone tool that aims to re-register
Magma's gateways in kubernetes cluster. It listens
kubernetes events and re-register gateways in orc8r.

### CODE REVIEW

File: orc8r/tools/magma-manipulator/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Magma manipulator
+This is standalone tool that helps automatically register [Magma](https://github.com/facebookincubator/magma) gateways.
+It works in Kubernetes cluster with virtlet only.
+For example if some pod with gateway was recreated by Kubernetes this tool
+will re-register the new gateway automatically in [Magma](https://github.com/facebookincubator/magma) orc8r.
+
+## Installation
+```
+git clone git@github.com:vladiskuz/magma-manipulator.git
```
Comment (mpgermano) [inline orc8r/tools/magma-manipulator/README.md:9]: Can you update this link to the magma repo and change the command to `cd magma/or8cr/tools/magma-manipulator`

File: orc8r/tools/magma-manipulator/magma_manipulator/config.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+k8s:
+    kubeconfig_path: /root/vkuzmin/test/magma-manipulator/magma_manipulator/kconfig
```
Comment (mpgermano) [inline orc8r/tools/magma-manipulator/magma_manipulator/config.yml:2]: Can you remove/change `vkuzmin` to something more generic in this config

File: orc8r/tools/magma-manipulator/magma_manipulator/magma_api.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+
+
+def _get_register_gateway_url(gw_net_type, gw_net):
+    if gw_net_type == 'carrier_wifi_network':
+        return 'magma/v1/cwf/{gw_net}/gateways'.format(gw_net=gw_net)
+    elif gw_net_type == 'feg':
+        return 'magma/v1/feg/{gw_net}/gateways'.format(gw_net=gw_net)
+
+
+def _get_register_gateway_data(gw_net_type, gw_id, gw_uuid, gw_key, gw_name):
```
Comment (mpgermano) [inline orc8r/tools/magma-manipulator/magma_manipulator/magma_api.py:75]: We should be able to get rid of this entire function by just registering the gateway with an existing config. See my comment below.

File: orc8r/tools/magma-manipulator/magma_manipulator/main.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+                magma_api.register_gateway(CONF.orc8r_api_url,
+                                           gw.network, gw.network_type,
+                                           gw.id, gw_uuid, gw_key,
+                                           gw.name, CONF.magma_certs_path)
+
+                magma_api.apply_gateway_config(CONF.orc8r_api_url,
+                                               gw.network, gw.network_type,
+                                               gw.id, gw.get_config(),
+                                               CONF.magma_certs_path)
```
Comment (mpgermano) [inline orc8r/tools/magma-manipulator/magma_manipulator/main.py:164]: We should just combine the `magma_api.register_gateway` and `magma_api.apply_gateway_config` calls. Any data that can be specified as configuration can instead be provided when we register the gateway. This helps ensure that we don't get into a bad state where registration succeeds, but the config update fails.

File: orc8r/tools/magma-manipulator/magma_manipulator/magma_api.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+                  gw_net_type=gw_net_type,
+                  gw_net=gw_net,
+                  gw_uuid=gw_uuid,
+                  gw_key=gw_key)
+    LOG.info(msg)
+
+    register_gw_url = _get_register_gateway_url(gw_net_type, gw_net)
+    magma_gw_url = urljoin(orc8r_api_url, register_gw_url)
+
+    data = _get_register_gateway_data(gw_net_type,
```
Comment (mpgermano) [inline orc8r/tools/magma-manipulator/magma_manipulator/magma_api.py:288]: Replace this call with a call to fetch previous configuration.

**Review by mpgermano — CHANGES_REQUESTED** [review]
Thanks for this! I just had a few comments to combine the registration and configuration steps.

**Comment by xjtian** [comment]
Thanks for this @vladiskuz! This is a good tool for our CWF deployments which run the cwag on k8s, but it's not the right time to pull something like this into trunk magma since we're focused on the fixed wireless story for now (i.e. AGW on baremetal). 

I think it'd be great to keep this tool opensourced under its own repository, which I see you've already done at https://github.com/vladiskuz/magma-manipulator. We can revisit the possibility of pulling this into trunk if our recommended deployment form factor for AGW shifts to k8s.

**Comment by zer0tweets** [comment]
Jacky, hey! Can you share more with us re: what you mean about "focus on fixed wireless story" ? Related, is there some place where we can find roadmap / list of near term feature priorities for Magma to make sure our upstream work is aligned with mainline direction?

**Pull request closed without merging**